packages/devtools/src/schema/resolution.ts

```diff
@@ -0,0 +1,340 @@
+    )
+  }
+  for (const model of dependencies) {
+    assertNoCircularDependency(models, targetModel, model, [...visited, currentModel])
+  }
+}
+
+export async function createIntermediaryCompositeDefinition(
+  ceramic: CeramicApi,
+  models: Record<string, AbstractModelDefinition>,
+): Promise<IntermediaryCompositeDefinition> {
+  const toResolve: Record<string, ResolveModel> = {}
+  for (const [modelName, definition] of Object.entries(models)) {
+    if (definition.action === 'load') {
+      toResolve[modelName] = {
+        requiredDependencies: [],
+        viewDependencies: [],
+        name: modelName,
+        execute: executeLoadFactory(ceramic, modelName, definition),
+      }
+    } else if (definition.action === 'create') {
+      const isInterface = definition.model.version !== '1.0' && definition.model.interface
+      const requiredDependencies = new Set<string>()
+      const viewDependencies = new Set<string>()
+
+      if (definition.model.version !== '1.0') {
+        for (const dependency of definition.model.implements) {
+          requiredDependencies.add(dependency)
+        }
+      }
+
+      for (const relation of Object.values(definition.model.relations ?? {})) {
+        if (relation.type === 'document' && relation.model !== null) {
+          if (relation.model === modelName) {
+            throw new Error(`Unsupported self-reference relation on model ${modelName}`)
+          }
+          requiredDependencies.add(relation.model)
+        }
+      }
+
+      for (const view of Object.values(definition.model.views ?? {})) {
+        if (
+          (view.type === 'relationCountFrom' ||
+            view.type === 'relationFrom' ||
+            view.type === 'relationDocument') &&
+          view.model !== null
+        ) {
+          if (isInterface) {
+            // Views must be present in the model definition of interfaces
+            requiredDependencies.add(view.model)
+          } else {
+            viewDependencies.add(view.model)
+          }
+        }
+      }
+
+      toResolve[modelName] = {
+        requiredDependencies: Array.from(requiredDependencies),
+        viewDependencies: Array.from(viewDependencies),
+        name: modelName,
+        execute: executeCreateFactory(ceramic, modelName, definition),
+      }
+    } else {
+      // @ts-ignore unknown action
+      throw new Error(`Unsupported action: ${definition.action}`)
+    }
+  }
+
+  const steps: Array<Array<ResolveModel>> = [[]]
+  const remainingModels = new Set<string>()
+
+  // In the first pass, check for circular dependencies and add models with no dependencies to the first execution step
+  for (const [name, resolve] of Object.entries(toResolve)) {
+    assertNoCircularDependency(toResolve, name)
+    if (resolve.requiredDependencies.length === 0) {
+      steps[0].push(resolve)
+    } else {
+      remainingModels.add(name)
+    }
+  }
+
+  // Add steps while there are remaining models to resolve
```

Comment:
ok so this part requires that we already checked for circular dependencies, so it doesnt get stuck, right?  I wonder if there is a lighter weight check we could do rather than searching the whole tree for them, but anyway that's fine.

Reply:
Yes I haven't tried to optimize the lookups but as it's only needed when creating composites and the number of models is likely pretty low, I don't think it's going to be an issue.